File: apps/dsa-web/src/api/universe.ts
import apiClient from './index';

export type AShareUniverseItem = {
  code: string;
  name: string;
  exchange: string;
  canonical_code: string;
  board?: string | null;
  industry?: string | null;
  active: boolean;
};

export type AShareUniverseSearchResponse = {
  query: string;
  items: AShareUniverseItem[];
  total: number;
  offset: number;
  limit: number;
};

export type AShareUniverseStats = {
  total_count: number;
  total_including_inactive: number;
  sync_status: 'idle' | 'running' | 'cooldown' | string;
  cooldown_seconds: number;
  cooldown_remaining_seconds: number;
  next_available_at?: string | null;
  last_triggered_at?: string | null;
  last_completed_at?: string | null;
  last_success_at?: string | null;
  last_error?: string | null;
  last_report: Record<string, unknown>;
  manual_only: boolean;
  auto_sync_enabled: boolean;
};

export type AShareUniverseManualSyncResponse = {
  accepted: boolean;
  message: string;
  sync_status: string;
};

export const universeApi = {
  async search(params: { q?: string; limit?: number; offset?: number }): Promise<AShareUniverseSearchResponse> {
    const { data } = await apiClient.get<AShareUniverseSearchResponse>('/api/v1/universe/a-share/search', {
      params: {
        q: params.q ?? '',
        limit: params.limit ?? 50,
        offset: params.offset ?? 0,
      },
    });
    return data;
  },

  async getStats(): Promise<AShareUniverseStats> {
    const { data } = await apiClient.get<AShareUniverseStats>('/api/v1/universe/a-share/stats');
    return data;
  },

  async triggerManualSync(): Promise<AShareUniverseManualSyncResponse> {
    const { data } = await apiClient.post<AShareUniverseManualSyncResponse>('/api/v1/universe/a-share/sync');
    return data;
  },
};

File: api/v1/schemas/universe.py
# -*- coding: utf-8 -*-
"""Universe search API schemas."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class AShareUniverseItem(BaseModel):
    code: str = Field(..., description="6 位 A 股代码")
    name: str = Field(..., description="股票简称")
    exchange: str = Field(..., description="交易所 SH/SZ/BJ")
    canonical_code: str = Field(..., description="带后缀的标准代码，如 600519.SH")
    board: Optional[str] = Field(None, description="板块")
    industry: Optional[str] = Field(None, description="行业")
    active: bool = Field(True, description="是否仍在 universe 中")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "code": "600519",
            "name": "贵州茅台",
            "exchange": "SH",
            "canonical_code": "600519.SH",
            "board": "主板",
            "industry": "白酒",
            "active": True,
        }
    })


class AShareUniverseSearchResponse(BaseModel):
    query: str = Field(..., description="原始查询关键字")
    items: List[AShareUniverseItem] = Field(default_factory=list, description="匹配结果")
    total: int = Field(..., description="返回条数")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "query": "茅台",
            "items": [
                {
                    "code": "600519",
                    "name": "贵州茅台",
                    "exchange": "SH",
                    "canonical_code": "600519.SH",
                    "board": "主板",
                    "industry": "白酒",
                    "active": True,
                }
            ],
            "total": 1,
        }
    })
